### DISCUSSION TITLE
Add @Instantiator annotation to mark the canonical constructor

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#5786] [Design Proposals]
I propose adding a `@Instantiator` annotation that allows marking the canonical constructor i.e. the constructor which takes in all properties, which should be used for instantiation.

As a first step, I would like to support this on embeddables to infer a `EmbeddableInstantiator` from the order of arguments of the constructor. This will require some byte code data flow analysis, but I implemented that in the past and I know it's usually doable. We can consider adding `String[] value() default {}` to represent the ordered set of property names for cases when the analysis fails, but let's only do that if absolutely necessary.

This would also affect column ordering for embeddables that represent structs, just like records are handled: https://github.com/hibernate/hibernate-orm/pull/5783
Also see https://hibernate.atlassian.net/browse/HHH-15898

**Comment by gavinking** [comment]
> This will require some byte code data flow analysis

Oh gawd of course I underestimated this one when we talked.

**Reply by beikov** [reply]
Ok, so ditched the data flow analysis and went with the explicit `String[]` now.
